RHIDP-1890 Create documentation for OIDC (#226)
redhat-developer/red-hat-developers-documentation-rhdh: getting-started/proc-configuring-keycloak-with-oidc.adoc

[id="proc-configuring-keycloak-with-oidc_{context}"]

= Configuring Keycloak with the OIDC authentication provider

//You can use Keycloak to authenticate with {product-short} using the OIDC authentication provider. 

{product} includes an OIDC authentication provider that can authenticate users by using Keycloak.

[IMPORTANT]
====
The user that you create in Keycloak must also be available in the {product-short} catalog.
====
////
.Prerequistes
. You configured {product-short} with a custom config map and secret.

.Procedure
////

.Procedure
. In Keycloak, create a new realm, for example `RHDH`.
. Add a new user.

*Username*:: Username for the user, for example: `rhdhuser`
*Email*:: Email address of the user.
*First name*:: First name of the user.
*Last name*:: Last name of the user.
*Email verified*:: Toggle to *On*.

> . Enter the *Password* for the user account and toggle *Temporary* to *Off*.

Reviewer: Could you add more information for the user like click on 'Create' button, and on the next page, select the 'Credentials' tab. Then, click on 'Set Password'

> . Enter a *Password* for the user that you have created and toggle *Temporary* to *Off*.

Reviewer 2: ```suggestion
. Enter the *Password* for the user account and toggle *Temporary* to *Off*.
```

. Create a new Client ID, for example, `RHDH`.

*Client authentication*:: Toggle to *On*.
*Valid redirect URIs*:: Set to the OIDC handler URL, for example, `https://<RHDH_URL>/api/auth/oidc/handler/frame`.

. Navigate to the *Credentials* tab and copy the *Client secret*.

. Save the Client ID and the Client Secret for the next step.

. In {product-short}, add your Keycloak credentials in your {product-short} secrets.

> .. Edit your Developer Hub secrets, such as secrets-rhdh.

Reviewer 2: ```suggestion
.. Edit your {product-short} secrets, such as `secrets-rhdh`.
```

Is `secrets-rhdh` a file or a parameter? Maybe we should specify that for clarity (i.e. Edit your {product-short} secrets [file/parameter], such as...)

Author: Elsewhere in our docs we just refer to 'secrets', I believe it is a term commonly used in OCP without specifically saying how it is stored e.g. in a file or as part of a config map.

`AUTH_KEYCLOAK_CLIENT_ID`:: Enter the Client ID that you generated in Keycloak, such as `RHDH`.

`AUTH_KEYCLOAK_CLIENT_SECRET`:: Enter the Client Secret that you generated in Keycloak.

. Set up the OIDC authentication provider in your {product-short} custom configuration.
+
.. Edit your custom {product-short} config map, such as `app-config-rhdh`.
+
.. In the `app-config-rhdh.yaml` content, add the `oidc` provider configuration under the root `auth` configuration, and enable the `oidc` provider for sign-in:
+
.`app-config-rhdh.yaml` fragment
[source,yaml,subs="+quotes,+attributes"]
----
auth:
  environment: production
  providers:
    oidc:
      production:
      clientId: ${AUTH_KEYCLOAK_CLIENT_ID}
      clientSecret: ${AUTH_KEYCLOAK_CLIENT_SECRET}
      metadataUrl: ${KEYCLOAK_BASE_URL}/auth/realms/${KEYCLOAK_REALM}
      prompt: ${KEYCLOAK_PROMPT} # recommended to use auto
      ## Uncomment for additional configuration options
      #callbackUrl: ${KEYCLOAK_CALLBACK_URL}
      #tokenEndpointAuthMethod: ${KEYCLOAK_TOKEN_ENDPOINT_METHOD}
      #tokenSignedResponseAlg: ${KEYCLOAK_SIGNED_RESPONSE_ALG}
      #scope: ${KEYCLOAK_SCOPE}

signInPage: oidc
----

.Verification

> . The `backstage-developer-hub` deployment starts a pod with the updated configuration.

Reviewer: I think it's important to say that the application needs to be restarted

Author: Done.

. Add a new user.

*Username*:: Username for the user, for example: `rhdhuser`
*Email*:: Email address of the user.
*First name*:: First name of the user.
*Last name*:: Last name of the user.
*Email verified*:: Toggle to *On*.

. Click *Create*.
. Navigate to the *Credentals* tab.
. Click *Set password*.
. Enter the *Password* for the user account and toggle *Temporary* to *Off*.

. Create a new Client ID, for example, `RHDH`.

*Client authentication*:: Toggle to *On*.
*Valid redirect URIs*:: Set to the OIDC handler URL, for example, `https://<RHDH_URL>/api/auth/oidc/handler/frame`.

. Navigate to the *Credentials* tab and copy the *Client secret*.

. Save the Client ID and the Client Secret for the next step.

. In {product-short}, add your Keycloak credentials in your {product-short} secrets.

.. Edit your Developer Hub secrets, such as secrets-rhdh.

.. Add the following key/value pairs:

`AUTH_KEYCLOAK_CLIENT_ID`:: Enter the Client ID that you generated in Keycloak, such as `RHDH`.

`AUTH_KEYCLOAK_CLIENT_SECRET`:: Enter the Client Secret that you generated in Keycloak.

. Set up the OIDC authentication provider in your {product-short} custom configuration.
+

> .. Edit your custom {product-short} config map, such as `app-config-rhdh`.

Reviewer 2: ```suggestion
.. Edit your custom {product-short} ConfigMap, such as `app-config-rhdh`.
```
